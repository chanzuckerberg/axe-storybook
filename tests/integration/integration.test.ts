/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { exec } from 'child_process';
import * as fs from 'fs';
import path from 'path';

// Before running these integration tests, the following steps must be completed:
//
// 1. `yarn demo:setup` - only needs to be ran when initially setting up the repo, or when changing
//                        the demo app.
// 2. `yarn build` - automatically ran by the pretest:integration step.

it('outputs accessibility violation information for the demo app', (done) => {
  expect.assertions(3);

  exec('yarn --cwd demo storybook:axe-no-build', function (error, stdout, stderr) {
    const normalizedStdout = normalize(stdout);
    const normalizedStderr = normalize(stderr);
    expect(error!.code).toEqual(1);
    expect(normalizedStdout).toMatchSnapshot();
    expect(normalizedStderr).toMatchSnapshot();
    
    done();
  });

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore TypeScript thinks these tests are Mocha, not Jest. Until we can figure out how to
  // deconflict the global Mocha and Jest types, we'll manually ignore the error.
}, 120000);

it('outputs accessibility violation information and sarif files for the demo app', (done) => {
  expect.assertions(12);

  exec('yarn --cwd demo storybook:axe-no-build:sarif', function (error, stdout, stderr) {
    const normalizedStdout = normalize(stdout);
    const normalizedStderr = normalize(stderr);
    expect(error!.code).toEqual(1);
    expect(normalizedStdout).toMatchSnapshot();
    expect(normalizedStderr).toMatchSnapshot();
    compareSarifFilesToSnapshot();

    done();
  });

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore TypeScript thinks these tests are Mocha, not Jest. Until we can figure out how to
  // deconflict the global Mocha and Jest types, we'll manually ignore the error.
}, 120000);

/**
 * Remove items from a string that are specific to a test run or environment, such as timing
 * information and file-system paths. That way, we can snapshot test effectively.
 */
function normalize(input: string) {
  /** Test times reported by Mocha. For example, `(520ms)` or `(3s)` */
  const specTimePattern = /\(\d+m?s\)/g;
  /** File system paths. For example, `/path/to/some/file */
  const cwdPattern = new RegExp(process.cwd(), 'g');
  /** Line numbers from stack trace paths. For example, `.js:20:55` */
  const lineNumbersPattern = /\.js:\d+:\d+/g;
  /** build/Suite.js is a path that is contained in the snapshot. Windows uses an absolute path. */
  const stackFilePathPattern = /(at Context\.<anonymous> \()([a-zA-Z0-9/\\\-.:]*?)([/\\]{1,4}build[/\\]{1,4}Suite.js\))/g;
  /** Stripping environment specific filepath */
  // const environmentFilePath = /: "file:\/\/\/.*?(\/axe-storybook-testing\/)/g;
  /** ISO DateTime. For example, `1970-01-01T00:00:00.0Z` */
  const isoDateTimePattern =/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/g;
  /** Ticks differ betwen Windows and other platforms */
  const tickPattern = /√/g;

  return input
    .replace(specTimePattern, '')
    .replace(cwdPattern, '')
    .replace(lineNumbersPattern, '.js')
    .replace(stackFilePathPattern, 'at Context.<anonymous> (/build/Suite.js)')
    // .replace(environmentFilePath, ': "file:///..$1')
    .replace(tickPattern, '✓')
    .replace(isoDateTimePattern, '1970-01-01T00:00:00.0Z');
}

/**
 * Compare outputted Sarif files to pre-determined snapshots.
 */
function compareSarifFilesToSnapshot() {
  const fileNames : string[] = [
    'button--button-1.sarif',
    'button--button-2.sarif',
    'button--button-3.sarif',
    'button--button-4.sarif',
    'button--button-6.sarif',
    'input--input-1.sarif',
    'input--input-2.sarif',
    'input--input-4.sarif',
    'input--input-5.sarif',
  ];
  
  fileNames.forEach(fileName => {
    let sarifFileContents : string = fs.readFileSync(
      path.join('./demo/sarif-test-results/', fileName),
      'utf8',
    );
    sarifFileContents = normalize(sarifFileContents);

    let sarifSnapshotFileContents : string = fs.readFileSync(
      path.join('./tests/integration/sarif-test-results-snapshots/', fileName),
      'utf8',
    );
    sarifSnapshotFileContents = normalize(sarifSnapshotFileContents);

    expect(sarifFileContents).toEqual(sarifSnapshotFileContents);
  });
}