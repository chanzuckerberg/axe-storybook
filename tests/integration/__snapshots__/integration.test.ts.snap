// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`outputs accessibility violation information for the demo app 1`] = `
"$ axe-storybook --output-sarif-files


  [chromium] accessibility
    button
      ✓ Button with text (should pass) 
      ✓ Button with emoji and label (should pass) 
      1) Button with no discernible text (should fail)
      2) Button with insufficient color contrast (should fail)
      - Button with insufficient color contrast but skipped (should pass)
      3) Multiple buttons with insufficient color contrast (should fail)
    input
      ✓ Input with label (should pass) 
      4) Input without label (should fail)
      - Input without label but skipped (should pass)
      5) Input without label and invalid role (should fail)
      ✓ Input without label but \\"label\\" rule is disabled (should pass) 


  4 passing 
  2 pending
  5 failing

  1) [chromium] accessibility
       button
         Button with no discernible text (should fail):
     AssertionError [ERR_ASSERTION]: 
Detected the following accessibility violations!

1. button-name (Buttons must have discernible text)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/button-name?application=axeAPI.

   Check these nodes:

   - html: <button style=\\"height: 25px; width: 50px;\\"></button>
     summary: Fix any of the following:
                Element does not have inner text that is visible to screen readers
                aria-label attribute does not exist or is empty
                aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                Element's default semantics were not overridden with role=\\"presentation\\"
                Element's default semantics were not overridden with role=\\"none\\"
                Element has no title attribute or the title attribute is empty

      at Context.<anonymous> (/build/Suite.js)

  2) [chromium] accessibility
       button
         Button with insufficient color contrast (should fail):
     AssertionError [ERR_ASSERTION]: 
Detected the following accessibility violations!

1. color-contrast (Elements must have sufficient color contrast)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/color-contrast?application=axeAPI.

   Check these nodes:

   - html: <button style=\\"background-color: red; color: hotpink;\\">Click me!</button>
     summary: Fix any of the following:
                Element has insufficient color contrast of 1.51 (foreground color: #ff69b4, background color: #ff0000, font size: 10.0pt (13.3333px), font weight: normal). Expected contrast ratio of 4.5:1

      at Context.<anonymous> (/build/Suite.js)

  3) [chromium] accessibility
       button
         Multiple buttons with insufficient color contrast (should fail):
     AssertionError [ERR_ASSERTION]: 
Detected the following accessibility violations!

1. color-contrast (Elements must have sufficient color contrast)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/color-contrast?application=axeAPI.

   Check these nodes:

   - html: <button style=\\"background-color: limegreen; color: hotpink;\\">Click me!</button>
     summary: Fix any of the following:
                Element has insufficient color contrast of 1.24 (foreground color: #ff69b4, background color: #32cd32, font size: 10.0pt (13.3333px), font weight: normal). Expected contrast ratio of 4.5:1

   - html: <button style=\\"background-color: hotpink; color: limegreen;\\">Click me!</button>
     summary: Fix any of the following:
                Element has insufficient color contrast of 1.24 (foreground color: #32cd32, background color: #ff69b4, font size: 10.0pt (13.3333px), font weight: normal). Expected contrast ratio of 4.5:1

      at Context.<anonymous> (/build/Suite.js)

  4) [chromium] accessibility
       input
         Input without label (should fail):
     AssertionError [ERR_ASSERTION]: 
Detected the following accessibility violations!

1. label (Form elements must have labels)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/label?application=axeAPI.

   Check these nodes:

   - html: <input placeholder=\\"Favorite coffee\\">
     summary: Fix any of the following:
                aria-label attribute does not exist or is empty
                aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                Form element does not have an implicit (wrapped) <label>
                Form element does not have an explicit <label>
                Element has no title attribute or the title attribute is empty
                Element's default semantics were not overridden with role=\\"none\\"
                Element's default semantics were not overridden with role=\\"presentation\\"

      at Context.<anonymous> (/build/Suite.js)

  5) [chromium] accessibility
       input
         Input without label and invalid role (should fail):
     AssertionError [ERR_ASSERTION]: 
Detected the following accessibility violations!

1. aria-roles (ARIA roles used must conform to valid values)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/aria-roles?application=axeAPI.

   Check these nodes:

   - html: <input placeholder=\\"Favorite coffee\\" role=\\"wut-the-wut\\">
     summary: Fix all of the following:
                Role must be one of the valid ARIA roles: wut-the-wut

2. label (Form elements must have labels)

   For more info, visit https://dequeuniversity.com/rules/axe/4.0/label?application=axeAPI.

   Check these nodes:

   - html: <input placeholder=\\"Favorite coffee\\" role=\\"wut-the-wut\\">
     summary: Fix any of the following:
                aria-label attribute does not exist or is empty
                aria-labelledby attribute does not exist, references elements that do not exist or references elements that are empty
                Form element does not have an implicit (wrapped) <label>
                Form element does not have an explicit <label>
                Element has no title attribute or the title attribute is empty
                Element's default semantics were not overridden with role=\\"none\\"
                Element's default semantics were not overridden with role=\\"presentation\\"

      at Context.<anonymous> (/build/Suite.js)



info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
"
`;

exports[`outputs accessibility violation information for the demo app 2`] = `
"error Command failed with exit code 1.
"
`;
